Attempt by Patrik Bóna

mrhead opened the pull request:
I am afraid that I am not able to do more refactoring, so here is my pull request.

mrhead committed:
Add rspec to project

* create Gemfile and add rspec gem
* create empty spec dir
* set some default options in .rspec

mrhead committed:
Add possibility to download file from URI

* URI content is stored in @file_content

mrhead committed:
Add XmlParser and possibility to count lines by speaker

mrhead committed:
Add possibility to retrieve all speakers

* with speaker method in XmlParser object

mrhead committed:
ShakespeareAnalyzer can print correct output parsed from XML file

mrhead committed:
Replace net/http with open-uri and add proxy support

mrhead committed:
Change method how to find all speakers

Declared PERSONAS don't mach SPEAKER in SPEECHEs, for example

<PERSONA>Three Witches.</PERSONA>
vs.
<SPEAKER>First Witch</SPEAKER>
...

whole xml: http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml

mrhead committed:
shakespeare_analyzer.rb can be now run as script

* run it with `ruby lib/shakespeare_analyzer.rb`

mrhead committed:
Add forgotten file from previous commit

mrhead committed:
Add script output to repository

Analyzed file:
http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml

mrhead committed:
Extract xml_parser method

mrhead committed:
Remove baseless attr_accessor for @file_content

mrhead committed:
Add xml_parser caching

mrhead committed:
Refactor XmlParser

* refactor `speakers` method to one line
* rename `lines_by_speaker` to `count_lines_by_speaker`

mrhead committed:
Add speeches with more speakers to XML test file

mrhead committed:
Extract `speakers` method in `ShakespeareAnalyzer`

Almost all methods are now one line long.

twosevenzero commented on output.txt:
```diff
@@ -0,0 +1,41 @@
+984 MACBETH
```
Even after fixing the bug in my code, I am not getting a number this high for Macbeth. Do we know what the actual answer is?

mrhead replied:
Well I do not know what are the correct numbers ;).

I perform my tests on small test.xml file and for this file everything is calculated correctly. But I do not know if I do not miss some edge cases in XML.

mrhead replied:
Well I just found bug in my code, I need to fix it. Hopefully then we will have same numbers ;).

mrhead replied:
Bug fixed, now we have same numbers for Macbeth ;)

twosevenzero commented on spec/shakespeare_analyzer_spec.rb:
```diff
@@ -0,0 +1,30 @@
+describe ShakespeareAnalyzer do
+  describe '#initialize' do
+    it 'reads provided URI and stores its content to @file_content' do
+      stub_request(:get, 'http://www.example.com/test_file.txt').to_return(body: 'This is just a test file!')
```
I need to learn what these stubs are. Do they interrupt flow and substitute content?

mrhead replied:
Yeah, check this gem: https://github.com/bblimke/webmock

I've found it when I was looking for solution how to test http download. It is awesome.

twosevenzero commented on spec/shakespeare_analyzer_spec.rb:
```diff
@@ -0,0 +1,30 @@
+    it 'reads provided URI and stores its content to @file_content' do
+      stub_request(:get, 'http://www.example.com/test_file.txt').to_return(body: 'This is just a test file!')
+      analyzer = ShakespeareAnalyzer.new('http://www.example.com/test_file.txt')
+      expect(analyzer.instance_variable_get(:@file_content)).to eq 'This is just a test file!'
```
I really could have used this method when I was writing my tests. "instance_variable_get"
Ill have to remember that for next time.

andyw8 replied:
I would actually recommend against using `instance_variable_get` in a test, because it means you're inspecting the internal state of the object, rather than testing it's external behaviour.

mrhead committed:
Fix bug in `count_lines_by_speaker` method

`count_lines_by_speaker` is now looking for exact match in speaker's
name.

Example XML:

<SPEECH>
  <SPEAKER>A</SPEAKER>
  <LINE>Line!</LINE>
</SPEECH>
<SPEECH>
  <SPEAKER>AAA</SPEAKER>
  <LINE>Line!</LINE>
</SPEECH>

before:
count_lines_by_speaker('A') == 2

now:
count_lines_by_speaker('A') == 1

mrhead committed:
Update output.txt (for MACBETH)

JESii commented:
Hi, Patrik... Nice work! You might want to consider updating the output so that the names aren't upper-cased.

mrhead committed:
Extract part for running as script from lib/ to bin/

mrhead committed:
Move logic from shakespeare_analyzer class to xml_parser class

mrhead committed:
Add missing test for XmlParser#initialize

andyw8 commented on bin/shakespeare_analyzer.rb:
```diff
@@ -0,0 +1,8 @@
+
+DEFAULT_XML_URL = 'http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml'
+
+analyzer = ShakespeareAnalyzer.new(DEFAULT_XML_URL)
```
I think it would be good to avoid passing a URL to the initializer, and just provide the data instead. That means you could supply text from a file, and it should also make testing easier.

andyw8 commented on lib/shakespeare_analyzer.rb:
```diff
@@ -0,0 +1,28 @@
+  private
+
+  def get_content_from_uri(uri)
+    open(uri, proxy: ENV['http_proxy']).read
```
This suggests the Shakespeare class is doing too much - it's parsing the text, but also having to deal with the complexities of a proxy server.

r00k replied:
Agreed with @andyw8.

andyw8 commented on spec/test_files/test.xml:
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0"?>
```
I like this concept of having the XML samples in a separate file. I embedded them within my spec, but it makes it very cluttered. I guess there's a trade-off though, in that you need to look at two places to understand what the spec is doing.

mrhead commented:
@andyw8 thanks for all your comments. They all seems valid to me and I will keep them in mind for my next projects. Thanks!

r00k commented on bin/shakespeare_analyzer.rb:
```diff
@@ -0,0 +1,8 @@
+
+require_relative '../lib/shakespeare_analyzer.rb'
+
+DEFAULT_XML_URL = 'http://www.ibiblio.org/xml/examples/shakespeare/macbeth.xml'
```
You're one of the only folks to extract this into a constant. I like it. :)

r00k commented on lib/shakespeare_analyzer.rb:
```diff
@@ -0,0 +1,28 @@
+  end
+
+  def run
+    print_speakers_sorted_by_line_count
```
Nice high level of abstraction.

r00k commented on spec/shakespeare_analyzer_spec.rb:
```diff
@@ -0,0 +1,33 @@
+  describe '#run' do
+    it 'prints list of speakers sorted by line count' do
+      test_file = File.dirname(__FILE__) + '/test_files/test.xml'
+      stub_request(:get, 'http://www.example.com/test.xml').to_return(body: File.read(test_file))
```
Notice how you have to stub out the http library when what you really want to do is check that the right parsing/counting is happening. This is a strong hint that your class has a few too many responsibilities.

charlieanna replied:
I never thought of stubbing the internet here. Thanks. Will add it to my solution too.

shreve commented on lib/shakespeare_analyzer.rb:
```diff
@@ -0,0 +1,28 @@
+    open(uri, proxy: ENV['http_proxy']).read
+  end
+
+  def print_speakers_sorted_by_line_count
```
Nice job keeping even your private methods expressive, readable, and short. This whole file is really easy to understand, but I particularly like the clarity of this method.
